### PR TITLE
De-biasing and variance estimates

### PR DESCRIPTION
**Pull request by byolkim**
adding methods for de-biasing (symmetric) KLIEP, and for computing the variance of (symmetric) KLIEP

### CODE REVIEW

File: src/utils_single_edge.jl [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    Ψx::Matrix{Float64},
+    Ψy::Matrix{Float64},)
+
+    μx = vec(mean(Ψx, dims=2))
+    ry = zeros(size(Ψy,2))
+    mul!(ry, transpose(Ψy), θ)
+
+    θ1 = θ[ind]
+    for k in ω.nzind
+        θ1 += ω[k] * (μx[k] - mean( exp.(ry) .* Ψy[k, :] ) / mean(exp, ry))
```
Comment (mlakolar) [inline src/utils_single_edge.jl]: You are recomputing 
`exp.(ry)
`and
`mean(exp, ry)`
in each iteration even though they do not change from iteration to iteration.

File: src/utils_single_edge.jl [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    rx = zeros(size(Ψx,2))
+    mul!(rx, transpose(Ψx), θ)
+    rx .*= -1.
+
+    ry = zeros(size(Ψy,2))
+    mul!(ry, transpose(Ψy), θ)
+
+    θ1 = θ[ind]
+    for k in ω.nzind
+        θ1 += ω[k] * (μx[k] - μy[k] + mean( exp.(rx) .* Ψx[k, :] ) / mean(exp, rx) - mean( exp.(ry) .* Ψy[k, :] ) / mean(exp, ry))
```
Comment (mlakolar) [inline src/utils_single_edge.jl]: Same here

File: src/utils_single_edge.jl [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    wy = zeros(ny)
+    mul!(wy, transpose(Ψy), θ)
+    wy .= exp.(wy)
+    wy ./= mean(wy)
+
+    # compute scaled data
+    Ψxsub = Ψx[supp, :]
+    Ψysub = Ψy[supp, :]
+    for k = 1:s
+        Ψysub[k, :] *= wy[k]
```
Comment (mlakolar) [inline src/utils_single_edge.jl]: Can you check if this is allocating additional memory?

File: src/utils_single_edge.jl [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    # compute scalars
+    wy = zeros(ny)
+    mul!(wy, transpose(Ψy), θ)
+    wy .= exp.(wy)
+    wy ./= mean(wy)
+
+    # compute scaled data
+    Ψyw = zeros(p,ny)
+    for k = 1:p
+        Ψyw[k, :] = wy[k] * Ψy[k, :]
```
Comment (mlakolar) [inline src/utils_single_edge.jl]: This is definitely allocating. Create another loop.

File: src/utils_single_edge.jl [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    wy = transpose(Ψy) * θ
+    wy .= exp.(wy)
+    wy ./= mean(wy)
+    wy .+= 1.
+
+    # compute scaled data
+    Ψxw = Ψx[supp, :]
+    for k = 1:s
+        Ψxw[k, :] *= wy[k]
```
Comment (mlakolar) [inline src/utils_single_edge.jl]: Check if allocating

File: src/utils_single_edge.jl [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    end
+
+    Ψyw = Ψy[supp, :]
+    for k = 1:s
+        Ψyw[k, :] *= wy[k]
```
Comment (mlakolar) [inline src/utils_single_edge.jl]: Check if allocating

File: src/utils_single_edge.jl [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+    wy = transpose(Ψy) * θ
+    wy .= exp.(wy)
+    wy ./= mean(wy)
+    wy .+= 1.
+
+    # compute scaled data
+    Ψxw = zeros(p,nx)
+    for k = 1:p
+        Ψxw[k, :] = Ψx[k, :] * wx[k]
```
Comment (mlakolar) [inline src/utils_single_edge.jl]: Rewrite so that it does not allocate.

File: src/utils_single_edge.jl [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    end
+
+    Ψyw = zeros(p,ny)
+    for k = 1:p
+        Ψyw[k, :] = Ψy[k, :] * wy[k]
```
Comment (mlakolar) [inline src/utils_single_edge.jl]: Rewrite

**Review by mlakolar — CHANGES_REQUESTED** [review]

File: src/utils_single_edge.jl (hunk 3 above)
Comment (byolkim) [reply src/utils_single_edge.jl]: Just discovered a bug. Lines 70-73 should be

    Ψysub = Ψy[supp, :]
    for k = 1:s
        Ψysub[k, :] .*= wy
    end

I suppose you are asking me to compare to

    Ψxsub = Ψx[supp, :]
    Ψysub = Ψy[supp, :]
    for k = 1:s
        for j = 1:ny
            Ψysub[k, j] *= wy[j]
        end
    end

Here's a comparison

julia> @time KLIEP_var(Ψx, Ψy, θhat1, ω)
  0.000028 seconds (26 allocations: 23.203 KiB)
0.006137912837101957

julia> @time KLIEP_var2(Ψx, Ψy, θhat1, ω)
  0.000028 seconds (24 allocations: 18.203 KiB)
0.006137912837101957

var2 is the changed version. Should I switch to a double loop?

File: src/utils_single_edge.jl (hunk 4 above)
Comment (byolkim) [reply src/utils_single_edge.jl]: The same bug is lurking here as well. Lines 108 - should read

    for k = 1:p
        Ψyw[k, :] = wy .* Ψy[k, :]
    end

Is the suggested fix

    for k = 1:p
        for j = 1:ny
            Ψyw[k, j] = wy[j] * Ψy[k, j]
        end
    end

Here's the comparison

julia> @time KLIEP_var(Ψx[supp3,:], Ψy[supp3,:], θtilde.nzval, ω)
  0.000040 seconds (28 allocations: 33.000 KiB)
0.012587931136663767

julia> @time KLIEP_var2(Ψx[supp3,:], Ψy[supp3,:], θtilde.nzval, ω)
  0.000042 seconds (24 allocations: 23.000 KiB)
0.012587931136663767

File: src/utils_single_edge.jl (hunk 1 above)
Comment (byolkim) [reply src/utils_single_edge.jl]: Fixed!

**Review by mlakolar — APPROVED** [review]
looks good

**Pull request merged**